webservice step 13

marcghaly opened the pull request:
Hi @keevan,

I am aware there is still an issue with the state but have not been able to troubleshoot it, maybe another pair of eyes can be helpful.

Resolves #13

Regards,

Marc

marcghaly commented:
Managed to trouble shoot state issue,

Regards,

Marc

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class flow_web_service extends flow_logic_step {
+    /**
```
Can you please add a blank line above this function's phpdoc

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+                ]
+            ]
+        ];
+        $ex['yaml'] = '<pre>users:
```
I'm not sure what this variable means just from looking at it.

May you consider changing this to something that is more meaningful / self explanatory and searchable?

https://github.com/jupeter/clean-code-php#use-meaningful-and-pronounceable-variable-names

marcghaly replied:
Changed $examples

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+            ]
+        ];
+        $ex['yaml'] = '<pre>users:
+        0:
+          username: john1234
+          password: Pwdtest123*&12
+          firstname: john
+          lastname: doe
+          email: john@doe.ca
+        </pre>';
+        $ex['json'] = \html_writer::nonempty_tag('pre', json_encode($jsonexample, JSON_PRETTY_PRINT));
```
You don't need to recreate the example in various different formats. I suggest keeping it in a PHP array, and using the methods to encode it to YAML and JSON as required. Possibly just showing the preferred format as an example, but listing all the supported formats.

marcghaly replied:
Fixed.

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+          firstname: john
+          lastname: doe
+          email: john@doe.ca
+        </pre>';
```
Can you please use the `html_writer::tag` method here instead? 

https://docs.moodle.org/dev/html_writer#References

marcghaly replied:
Fixed.

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+     */
+    protected static function form_define_fields(): array {
+        return [
+            'webservice' => ['type' => PARAM_RAW_TRIMMED],
+            'user' => ['type' => PARAM_TEXT],
+            'parameters' => ['type' => PARAM_RAW],
+            'datastore' => ['type' => PARAM_RAW_TRIMMED],
+            'failure' => ['type' => PARAM_TEXT],
+            'path' => ['type' => PARAM_PATH],
```
The fields now support a new type (bool) called 'required', so adding this should automatically apply a form rule to the relevant field. This should currently be used in addition to the validate_config method, because of the different _forms_ of validation. 

And would have otherwise been asked to be included in the custom form inputs below if not already.

Ref:
https://docs.moodle.org/dev/lib/formslib.php_Form_Definition#addRule

marcghaly replied:
Fixed.

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+     *
+     * @return array $return.
+     */
+    public function array_evaluate_recursive($arr, $input, $expressionlanguage) {
+        foreach ($arr as $key => $value) {
+            if (is_array($value)) {
+                $value = $this->array_evaluate_recursive($value, $input, $expressionlanguage);
+                $return[$key] = $value;
+            } else {
+                end($arr);
+                $lastkey = key($arr);
+                if (strpos($value, 'data.') !== false) {
+                    $res = $expressionlanguage->evaluate($value, ['data' => $input]);
+                    $arr[$key] = strtolower($res);
+                }
+                if ($key == $lastkey) {
+                    $return = $arr;
+                }
+            }
+        }
+        return $return;
+    }
```
If my understanding is correct, this does the same thing the parser would, but this supports an array coming in, and a processed array coming back out (and some hard rules for the data field)

I think this should be moved into the parser class, and it should use the evaluation function of the parser class instead of the raw expression language class for the following reasons:
- Keeps usage of expression variables the same throughout the dataflows plugin
- It will actually support setting webservice fields to something that can be referenced in an expression. I believe currently you would be bound to whatever $data you pass through.
- Keeps how it's used centralized in one place, which helps with maintainability.

This will mean you'll need to update usage to pass in the dataflow variables as well as the $input fields under e.g. a data key.

And this will also mean the configuration you've set currently should probably be `${{ data.FIRST_NAME }}` instead of `data.FIRST_NAME`

marcghaly replied:
I have made changes pertaining to the parse in a separate commit, I m opening another PR that should be merged before with this, if all other changed are ok I ll drop the other commit.

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+                    file_put_contents($destination, $response['exception']);
+                    return false;
+                }
+            } else {
```
Unnecessary else / nested block. 

Reason: The if guard used above for `$response['error']` is expected to throw / return / halt execution. So anything after that should be expected to be working and normal. 

https://github.com/jupeter/clean-code-php#avoid-nesting-too-deeply-and-return-early-part-1

marcghaly replied:
Fixed. Thank you

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+            } else {
+                // Success - store any desired value for subsequent steps.
+                if (!empty($config->datastore)) {
+                    $data = explode(',', $config->datastore);
```
This variable name is a bit generic. Including usage of $data and $val below it's a bit hard to know what's going on.

May you consider changing this to something that is more meaningful / self explanatory and searchable?

https://github.com/jupeter/clean-code-php#use-meaningful-and-pronounceable-variable-names

marcghaly replied:
Fixed.

keevan commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+                // Success - store any desired value for subsequent steps.
+                if (!empty($config->datastore)) {
+                    $data = explode(',', $config->datastore);
+                    foreach ($data as $val) {
+                        $value = $this->array_search_key($response['data'], $val);
+                        $value = $value[$val];
+                        $val = $this->stepdef->alias . '_' . $val;
+                        $input->{$val} = $value;
+                    }
```
With regards to the usage here, I'm somewhat on the fence about how it's being used currently. Is it not possible to reference what you want directly? And how you want it passed on directly as well? I think this can be more generic, and could be applied across steps as needed.

It's somewhat related to this: https://github.com/catalyst/moodle-tool_dataflows/issues/260 but the idea I had in mind at least for this step initially is something like, the configuration can be in yaml format
```
userid: newuser.id
username: newuser.username
```

This is based on the assumption the result of the web service call being nested in a field called newuser, the userid and username fields would then be merged in with the $input value so the data would be "transformed" by this web service step. 

Perhaps that is one way to do it? I'm wary of value conflicts and setting aliases so this seems like a way to resolve that issue.

marcghaly replied:
This issue I ran through here is that Web services such as `core_user_create_users` and `core_course_get_courses_by_field` have responses that have `id` keys, by adding the alias it solves the conflicting values overriding each other. Values are being added to the input so they are reused by subsequent steps.

keevan reviewed (commented):
You will need to bump the version.php because:
- lang strings have been updated

I've done a quick code pass, but will test functionality soon.

keevan commented:
Could you also please add some unit tests

keevan commented:
Could you also add a "Fixes #13" or "Resolves #13" to the PR description. This links the PR to the issue and automatically closes it after this PR merges.

It also shows as a linked pull request from the project board. Example below:

![image](https://user-images.githubusercontent.com/9924643/179147673-a598f5a9-68c3-47c1-9218-2405811cfd90.png)

marcghaly commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class flow_web_service extends flow_logic_step {
+    /**
```
Added.

marcghaly commented on tests/tool_dataflows_web_service_flow_test.php:
```diff
@@ -0,0 +1,218 @@
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class tool_dataflows_web_service_flow_test extends \advanced_testcase {
+    /**
```
Added.

marcghaly commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+                    'firstname' => 'john',
+                    'lastname' => 'doe',
+                    'email' => 'john@doe.ca',
+                ]
+            ]
```
Added.

marcghaly commented on tests/tool_dataflows_web_service_flow_test.php:
```diff
@@ -0,0 +1,218 @@
+            'curl' => $this->getExternalTestFileUrl('/h5puuid.json'),
+            'destination' => 'test.html',
+            'headers' => '',
+            'method' => 'get'
```
Added.

marcghaly commented on tests/tool_dataflows_web_service_flow_test.php:
```diff
@@ -0,0 +1,218 @@
+                    'firstname' => 'john',
+                    'lastname' => 'doe',
+                    'email' => 'john@doe.ca',
+                ]
+            ]
```
Added.

marcghaly commented on tests/tool_dataflows_web_service_flow_test.php:
```diff
@@ -0,0 +1,218 @@
+            'curl' => $this->getExternalTestFileUrl('/h5pcontenttypes.json'),
+            'destination' => 'test1.html',
+            'headers' => '',
+            'method' => 'get'
```
Added.

marcghaly commented on tests/tool_dataflows_web_service_flow_test.php:
```diff
@@ -0,0 +1,218 @@
+                    'firstname' => 'data.owner',
+                    'lastname' => 'doe',
+                    'email' => 'john@dodoe.ca',
+                ]
+            ]
```
Added

marcghaly commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,281 @@
+            get_string('flow_web_service:field_parameters_help', 'tool_dataflows', $ex));
+        $mform->addElement('text', 'config_datastore', get_string('flow_web_service:datastore', 'tool_dataflows'));
+        $mform->addHelpButton('config_datastore', 'flow_web_service:datastore', 'tool_dataflows');
+        $mform->addElement('static', 'datastore_example', '', get_string('flow_web_service:datastore_example', 'tool_dataflows', $this->stepdef->alias));
```
fixed

marcghaly commented:
Hi @keevan,

I have used the outputs and made a few changes to parser / base_step, let me know if the use is incorrect,
Made appropriate changes to unit tests as well.

Regards,

Marc

marcghaly commented:
Hi @keevan, I have rebased over #309  and squashed those commits, if all is good to you I will rebase this branch over main branch once #309 is merged.

Regards,

Marc

brendanheywood commented:
@marcghaly I tried to resolve the merge conflicts in the github ui and it made a bit of a mess, can you please rebase this and push a clean patch up

Once more thing for in future (no need to retro do it here), can you make branch names which are human readable, eg if the issue was 13-webservice-step this makes it much easier when jumping around a lot of branches and cleaning them up later

brendanheywood commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,229 @@
+     * @param \MoodleQuickForm &$mform
+     */
+    public function form_add_custom_inputs(\MoodleQuickForm &$mform) {
+        $users = get_users(true, '', true, null, 'firstname ASC', '', '', '', 25, 'id, firstname, lastname, email');
```
I don't think this can work, most sites will have many thousands of users.

So can you make it so that if it is left blank then it looks up the primary user. If it is specified then it looks up that user and I think it can just be a plain text field an accept the username of the user. This should validate when saving. Make the help text and placeholder text clear what is going on

marcghaly replied:
Done.

brendanheywood commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,229 @@
+            $SESSION = $session;
+            $OUTPUT = $currentoutput;
+        } else {
+            // Restore the previous user to avoid any side-effects occuring in later steps / code.
```
This chunk is repeated 3 times, can you reorder things and refactor this into one?

marcghaly replied:
Fixed

brendanheywood commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,229 @@
+                }
+                if ($failure === 'record') {
+                    $destination = $this->enginestep->engine->resolve_path($path);
+                    file_put_contents($destination, $response['exception'] ?? $response['exception']->message);
```
If there are multiple errors for different rows this will overwrite it each time, if we want this to be in a file then we should append it. 

But I'm not even sure putting it into a file is the most useful though as it isn't easily accessible for a site admin to view and do anything with. I think for MVP lets just log it into the flow logs, we can come back to this in another issue when with some real world use cases

brendanheywood commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,229 @@
+                $OUTPUT = $currentoutput;
+
+                // Throw an exception to be propagated for proper error capture.
+                if (!$outertransaction && $DB->is_transaction_started()) {
```
Can you explain this bit in more detail around the transactions in the comment?

marcghaly replied:
This is just a layer of safety in case the Webservice needs to be writting and fails - just ensure it does not leave the db in an uncomplete state

brendanheywood replied:
I meant in the comment in code. But the bit I specifically don't understand is the handling for nested transactions

marcghaly commented:
Hi @brendanheywood, I rebased this branch over latest main one, there one comment I am not sure to have understood concerning file put contents, do you want this to be dropped or just leave it as is and fix it later ?

Regards,

Marc

brendanheywood commented:
Right now is crunch time, so if BCPS don't explicitly need a feature then lets err on the side of dropping the feature and splitting it into a new issue for later. So can you please do that here, just remove the option, make an issue, and you can stash the wip in another branch to come back to

brendanheywood commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,220 @@
+        }
+        if ($config->user) {
+            global $DB;
+            $userid = $DB->get_field('user', 'id', ['username' => $config->user]);
```
Can you also check that they are not suspended and not deleted. Even better use get_users() as it does some of this for you

marcghaly replied:
Done.

brendanheywood commented on lib.php:
```diff
@@ -61,6 +61,7 @@ function tool_dataflows_step_types() {
         new step\connector_s3,
         new step\connector_sns_notify,
         new step\connector_wait,
+        new step\flow_web_service,
```
can you please keep this list sorted by alpha (and fix any others which are wrong)

marcghaly replied:
Fixed.

marcghaly commented:
Hi @brendanheywood, I have made requested changes and opened another issue for logging the output to a file #368. Concerning the comment around transaction, it was a forgotten one, I have removed it as it is not throwing at this place.

Regards,

Marc

keevan reviewed (changes requested):
Some small things:
```
-------------
PHPCS issues
-------------
classes/local/step/flow_web_service.php:122:146: warning - Line exceeds 132 characters; contains 146 characters

--------------------------
VERSION NEEDS TO BE BUMPED
--------------------------
- lang strings have been updated
-------------
PHPDOC issues
-------------
admin/tool/dataflows/classes/local/step/flow_web_service.php
    Line 56: Phpdocs for function flow_web_service::form_add_custom_inputs has incomplete parameters list
```

Unit test is also failing, most likely because the user field has changed from using the `id` of the user and now expects the `username` instead. I also encountered this issue and had to change the previously provided dataflow.

The help text for the parameters field also needs to be updated, in particular it is now expecting `${{ data.something }}` instead of simply `data.something`, which allows for both forms if required. See screenshot below:

![image](https://user-images.githubusercontent.com/9924643/181666735-87c3a47d-6f5e-425c-9e83-782400f555a5.png)



Once the above things are fixed up I'm happy to merge this for now and catch any issues at a later stage, as functionally it seems to be working.

brendanheywood commented on classes/local/step/flow_web_service.php:
```diff
@@ -0,0 +1,213 @@
+
+    /**
+     * Return the definition of the fields available in this form.
+     *
```
Also this step should declare that it has side effects. This is not as simple as just saying true, it should really leverage the web service definition to say whether or not it is a read ws or not.

eg core_auth_confirm_user is a write ws, but core_auth_is_minor is a read one:

https://github.com/moodle/moodle/blob/master/lib/db/services.php#L49-L57

If the absence of config this step should by default be considered a write step which has side effects.

marcghaly replied:
Added that here https://github.com/catalyst/moodle-tool_dataflows/commit/66e6f6e453afaf7b5288bd4fc27388310659d48a#diff-aa35067838cac274f171b1a761e9e37cc4652fbd570662aa5e9c3dd7bdab2aa1R192, but unsure if that should be always false when dry run and if it should be in the has_side_effect method instead ?

marcghaly commented:
Hi @brendanheywood and @keevan, I have fixed issues above and rebased against main branch.

Regards,

Marc
